File: PythonSolutions/P03/HighestProduct.py
# Given an arrayOfInts, find the highestProduct you can get from three of the integers.
# The input arrayOfInts will always have at least three integers.
from heapq import *

def highestProductOf3(arr):
	smallest = []
	largest = []
	for element in arr:
		if len(largest) < 3:
			heappush(largest, element)
		else:
			heappushpop(largest, element)
		if len(smallest) < 2:
			heappush(smallest, -1 * element)
		else:
			heappushpop(smallest, -1 * element)
	cand1 = largest[0] * largest[1] * largest[2]
	cand2 = smallest[0] * smallest[1] * max(largest)
	return max(cand1, cand2)

def test(arr):
	print "Largest Product of Three in %s = %d" % (str(arr), highestProductOf3(arr))

arr = [1, 10, -5, 1, -100]
test(arr)
arr = [-10, -10, 1, 3, 2]
test(arr)
arr = [-10, -3, -2, -5, -6, -4, -8]
test(arr)
arr = [1, 4, 5, 2, 9, 10, 3]
test(arr)

File: JavaSolutions/P03/HighestProduct.java
/*  Given an arrayOfInts, find the highestProduct you can get 
    from three of the integers.
 */

import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.Comparator;
import java.lang.Math;

public class HighestProduct03 {
    public static int highestProd(int[] a) {
	int max = a[0] > a[1] ? a[0] : a[1]; // max so far
	int min = a[0] > a[1] ? a[1] : a[0]; // min so far
	int lp2 = max * min;                 // largest product of 2 so far
	int sp2 = lp2;                       // smallest product of 2 so far
	int lp3 = Integer.MIN_VALUE;         // largest product of 3 so far
	for (int i = 2; i < a.length; i++) {
	    int c = a[i];
	    lp3 = Math.max(c * lp2, Math.max(c * sp2, lp3));
	    lp2 = Math.max(c * max, Math.max(c * min, lp2));
	    sp2 = Math.min(c * min, Math.min(c * max, sp2));
	    if (max < c) max = c;
	    if (min > c) min = c;
	}
	return lp3;
    }

    public static void test(int[] a) {
	System.out.println("Input array:      " + Arrays.toString(a));
	System.out.println("Max product of 3: " + highestProd(a));
        System.out.println("_______________________________________");
    }

    public static void main(String[] args) {
	int[] a = new int[] {1, 2, 3, -10, -9, 8, 7, -2, 0};
	test(a);

	a = new int[] {-1, -2, -3, -4, -5, -6, -7};
	test(a);

	a = new int[] {-1, -2, -3, -4, -5, -6, -7, 0};
	test(a);

	a = new int[] {1, 2, 3, 5, -7, -2};
	test(a);

	a = new int[] {1, 2, 3, 5, -7};
	test(a);
    }
}
